src/Adapter/Product/Update/ProductSupplierUpdater.php

```diff
@@ -63,42 +69,103 @@ class ProductSupplierUpdater
 
     /**
      * @param ProductRepository $productRepository
+     * @param CombinationRepository $combinationRepository
      * @param SupplierRepository $supplierRepository
      * @param ProductSupplierRepository $productSupplierRepository
      */
     public function __construct(
         ProductRepository $productRepository,
+        CombinationRepository $combinationRepository,
         SupplierRepository $supplierRepository,
         ProductSupplierRepository $productSupplierRepository
     ) {
         $this->productRepository = $productRepository;
         $this->supplierRepository = $supplierRepository;
         $this->productSupplierRepository = $productSupplierRepository;
+        $this->combinationRepository = $combinationRepository;
     }
 
     /**
      * @param ProductId $productId
      * @param SupplierId $defaultSupplierId
      * @param ProductSupplier[] $productSuppliers
      *
-     * @throws CannotUpdateProductSupplierException
-     * @throws CoreException
-     * @throws ProductSupplierNotFoundException
+     * @return ProductSupplierId[]
      */
-    public function setProductSuppliers(ProductId $productId, SupplierId $defaultSupplierId, array $productSuppliers): void
-    {
-        $deletableProductSupplierIds = $this->getDeletableProductSupplierIds($productId->getValue(), $productSuppliers);
+    public function setProductSuppliers(
+        ProductId $productId,
+        SupplierId $defaultSupplierId,
+        array $productSuppliers
+    ): array {
+        $product = $this->productRepository->get($productId);
 
-        foreach ($productSuppliers as $productSupplier) {
-            if ($productSupplier->id) {
-                $this->productSupplierRepository->update($productSupplier);
-            } else {
-                $this->productSupplierRepository->add($productSupplier);
-            }
+        if ($product->cache_default_attribute) {
+            throw new RuntimeException(sprintf(
+                'Product #%d has combinations. Use %s::%s to set product suppliers for specified combination',
+                $productId->getValue(),
+                self::class,
+                'setCombinationSuppliers()'
+            ));
         }
 
-        $this->productSupplierRepository->bulkDelete($deletableProductSupplierIds);
+        $this->persistProductSuppliers($productId, $productSuppliers);
         $this->updateDefaultSupplier($productId, $defaultSupplierId);
+
+        return $this->getCurrentProductSupplierIds($productId);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param CombinationId $combinationId
+     * @param array<int, ProductSupplier> $productSuppliers
+     *
+     * @return ProductSupplierId[]
+     */
+    public function setCombinationSuppliers(
+        ProductId $productId,
+        CombinationId $combinationId,
+        array $productSuppliers
+    ): array {
+        $this->persistProductSuppliers($productId, $productSuppliers, $combinationId);
```

Comment:
If I remember right we discussed about an issue, when you switch from simple to combination product There may be some ProductSupplier left with combinationId equal to zero Shouldn't they be cleaned here?

(same question in `setProductSuppliers` maybe we should remove all former ProductSupplier associated to former combinations?)